docs/codeql/codeql-language-guides/customizing-library-models-for-java.rst

```diff
@@ -0,0 +1,439 @@
+
+The CodeQL library for Java exposes the following extensible predicates:
+
+- **sourceModel**\(package, type, subtypes, name, signature, ext, output, kind, provenance). This is used for **source** models.
+- **sinkModel**\(package, type, subtypes, name, signature, ext, input, kind, provenance). This is used for **sink** models.
+- **summaryModel**\(package, type, subtypes, name, signature, ext, input, output, kind, provenance). This is used for **summary** models.
+- **neutralModel**\(package, type, name, signature, provenance). This is used for **neutral** models, which does not impact the data flow analysis, but it is included here for completeness.
+
+The extensible predicates are populated using data extensions specified in YAML files.
+
+In the sections below, we will provide examples of how to add tuples to the different extensible predicates.
+The extensible predicates are used to customize and improve the existing data flow queries, by providing sources, sinks, and flow through (summaries) for library elements.
+The :ref:`reference-material` section will provide details on the *mini DSLs* that defines models for each extensible predicate.
+
+Example: Taint sink in the **java.sql** package.
+------------------------------------------------
+
+In this example we will show how to model the argument of the **execute** method as a SQL injection sink.
+This is the **execute** method in the **Statement** class, which is located in the **java.sql** package.
+Please note that this sink is already added to the CodeQL Java analysis.
+
+.. code-block:: java
+
+   public static void taintsink(Connection conn, String query) throws SQLException {
+       Statement stmt = conn.createStatement();
+       stmt.execute(query); // The argument to this method is a SQL injection sink.
+   }
+
+This means that we want to add a tuple to the **sinkModel**\(package, type, subtypes, name, signature, ext, input, kind, provenance) extensible predicate, which can be achieved by adding the following to a data extension file:
+
+.. code-block:: yaml
+
+   extensions:
+     - addsTo:
+         pack: codeql/java-all
+         extensible: sinkModel
+       data:
+         - ["java.sql", "Statement", True, "execute", "(String)", "", "Argument[0]", "sql", "manual"]
+
+Reasoning:
+
+Since we are adding a new sink, we need to add a tuple to the **sinkModel** extensible predicate.
+The first five values identify the callable (in this case a method) to be modeled as a sink.
+
+- The first value **java.sql** is the package name.
+- The second value **Statement** is the name of the class (type) that contains the method.
+- The third value **True** is a flag that indicates whether or not the sink also applies to all overrides of the method.
+- The fourth value **execute** is the method name.
+- The fifth value **(String)** is the method input type signature.
+
+The sixth value is only relevant internally and can be omitted in most use cases.
```

Comment:
It might be worth clarifying what you mean by "only relevant internally" but I may be betraying my layman status here so feel free to ignore this

Reply:
It is a good point. I will just write that it should be left empty and it is out of scope for this documentation.

Reply:
Sounds good 👍